File: python/assembler.py
from lexer import tokenize

def parse_def_args(tokens, pos):
    args = []
    if pos < len(tokens) and tokens[pos]["type"] == 'LPAREN':
        pos += 1
        while pos < len(tokens) and tokens[pos]["type"] != 'RPAREN':
            if tokens[pos]["type"] == 'SYMBOL':
                args.append(tokens[pos]["value"])
            pos += 1
            if pos < len(tokens) and tokens[pos]["type"] == 'COMMA':
                pos += 1
        if pos < len(tokens) and tokens[pos]["type"] == 'RPAREN':
            pos += 1
    return args, pos

def collect_call_args(tokens, pos, argc):
    args = []
    if pos < len(tokens) and tokens[pos]["type"] == 'LPAREN':
        pos += 1
        for _ in range(argc):
            arg_tokens = []
            while pos < len(tokens) and tokens[pos]["type"] != 'COMMA' and tokens[pos]["type"] != 'RPAREN':
                arg_tokens.append(tokens[pos])
                pos += 1
            args.append(arg_tokens)
            if pos < len(tokens) and tokens[pos]["type"] == 'COMMA':
                pos += 1
        if pos < len(tokens) and tokens[pos]["type"] == 'RPAREN':
            pos += 1
        return args, pos
    return None, pos

def substitute(body, params, args):
    result = []
    for body_token in body:
        substituted = False
        if body_token["type"] == 'SYMBOL':
            for idx, param_name in enumerate(params):
                if body_token["value"] == param_name:
                    if idx < len(args):
                        result.extend(args[idx])
                    substituted = True
                    break
        if not substituted:
            result.append(body_token)
    return result

def expand(tokens, macros, symbols, output, pc):
    i = 0
    while i < len(tokens):
        t = tokens[i]

        if t["type"] == 'HASH' and (i + 1) < len(tokens) and tokens[i + 1]["value"] == 'define':
            name = tokens[i + 2]["value"]
            params, next_pos = parse_def_args(tokens, i + 3)
            body = []
            while next_pos < len(tokens):
                bt = tokens[next_pos]
                if bt["type"] == 'SYMBOL' and bt["value"].lower() == 'endm':
                    next_pos += 1
                    break
                body.append(bt)
                next_pos += 1
            macros[name] = {"params": params, "body": body, "argc": len(params)}
            i = next_pos
        elif t["type"] == 'LABEL':
            symbols[t["value"]] = pc
            i += 1
        elif t["type"] == 'SYMBOL' and t["value"] in macros:
            macro = macros[t["value"]]
            args, next_pos = collect_call_args(tokens, i + 1, macro["argc"])
            if args is not None or macro["argc"] == 0:
                substituted = substitute(macro["body"], macro["params"], args or [])
                pc = expand(substituted, macros, symbols, output, pc)
                i = next_pos
            else:
                output.append(t)
                pc += 1
                i += 1
        elif t["type"] == 'NUMBER' or t["type"] == 'SYMBOL':
            output.append(t)
            pc += 1
            i += 1
        else:
            i += 1
    return pc

def pass1(source):
    tokens = tokenize(source) if isinstance(source, str) else source
    macros, symbols, expanded = {}, {}, []
    pc = expand(tokens, macros, symbols, expanded, 0)
    return {
        "macros": macros,
        "symbols": symbols,
        "tokens": expanded,
        "pc": pc
    }

def assemble(source):
    res = pass1(source)
    binary = []
    for t in res["tokens"]:
        if t["type"] == 'SYMBOL':
            addr = res["symbols"].get(t["value"])
            if addr is None:
                raise Exception(f"Undefined label: {t['value']}")
            binary.append(addr)
        elif t["type"] == 'NUMBER':
            binary.append(int(float(t["value"])))
    return binary

File: go/assembler.go
package subleq

import (
	"fmt"
	"strconv"
	"strings"
)

type Macro struct {
	Params []string `json:"params"`
	Body   []Token  `json:"body"`
	Argc   int      `json:"argc"`
}

type Pass1Result struct {
	Macros  map[string]Macro `json:"macros"`
	Symbols map[string]int   `json:"symbols"`
	Tokens  []Token          `json:"tokens"`
	PC      int              `json:"pc"`
}

func parseDefArgs(tokens []Token, pos int) ([]string, int) {
	var args []string
	if pos < len(tokens) && tokens[pos].Type == "LPAREN" {
		pos++
		for pos < len(tokens) && tokens[pos].Type != "RPAREN" {
			if tokens[pos].Type == "SYMBOL" {
				args = append(args, tokens[pos].Value)
			}
			pos++
			if pos < len(tokens) && tokens[pos].Type == "COMMA" {
				pos++
			}
		}
		if pos < len(tokens) && tokens[pos].Type == "RPAREN" {
			pos++
		}
	}
	return args, pos
}

func collectCallArgs(tokens []Token, pos int, argc int) ([][]Token, int) {
	var args [][]Token
	if pos < len(tokens) && tokens[pos].Type == "LPAREN" {
		pos++
		for a := 0; a < argc; a++ {
			var argTokens []Token
			for pos < len(tokens) && tokens[pos].Type != "COMMA" && tokens[pos].Type != "RPAREN" {
				argTokens = append(argTokens, tokens[pos])
				pos++
			}
			args = append(args, argTokens)
			if pos < len(tokens) && tokens[pos].Type == "COMMA" {
				pos++
			}
		}
		if pos < len(tokens) && tokens[pos].Type == "RPAREN" {
			pos++
		}
		return args, pos
	}
	return nil, pos
}

func substitute(body []Token, params []string, args [][]Token) []Token {
	var result []Token
	for _, bodyToken := range body {
		substituted := false
		if bodyToken.Type == "SYMBOL" {
			for idx, paramName := range params {
				if bodyToken.Value == paramName {
					if idx < len(args) {
						result = append(result, args[idx]...)
					}
					substituted = true
					break
				}
			}
		}
		if !substituted {
			result = append(result, bodyToken)
		}
	}
	return result
}

func expand(tokens []Token, macros map[string]Macro, symbols map[string]int, output *[]Token, pc int) (int, error) {
	i := 0
	for i < len(tokens) {
		t := tokens[i]

		if t.Type == "HASH" && (i+1) < len(tokens) && tokens[i+1].Value == "define" {
			name := tokens[i+2].Value
			params, nextPos := parseDefArgs(tokens, i+3)
			var body []Token
			for nextPos < len(tokens) {
				bt := tokens[nextPos]
				if bt.Type == "SYMBOL" && strings.ToLower(bt.Value) == "endm" {
					nextPos++
					break
				}
				body = append(body, bt)
				nextPos++
			}
			macros[name] = Macro{Params: params, Body: body, Argc: len(params)}
			i = nextPos
		} else if t.Type == "LABEL" {
			symbols[t.Value] = pc
			i++
		} else if t.Type == "SYMBOL" {
			if macro, ok := macros[t.Value]; ok {
				args, nextPos := collectCallArgs(tokens, i+1, macro.Argc)
				if args != nil || macro.Argc == 0 {
					substituted := substitute(macro.Body, macro.Params, args)
					var err error
					pc, err = expand(substituted, macros, symbols, output, pc)
					if err != nil {
						return 0, err
					}
					i = nextPos
				} else {
					*output = append(*output, t)
					pc++
					i++
				}
			} else {
				*output = append(*output, t)
				pc++
				i++
			}
		} else if t.Type == "NUMBER" {
			*output = append(*output, t)
			pc++
			i++
		} else {
			i++
		}
	}
	return pc, nil
}

func Pass1(source interface{}) (*Pass1Result, error) {
	var tokens []Token
	var err error
	if s, ok := source.(string); ok {
		tokens, err = Tokenize(s)
		if err != nil {
			return nil, err
		}
	} else if t, ok := source.([]Token); ok {
		tokens = t
	} else {
		return nil, fmt.Errorf("source must be string or []Token")
	}

	macros := make(map[string]Macro)
	symbols := make(map[string]int)
	var expanded []Token
	pc, err := expand(tokens, macros, symbols, &expanded, 0)
	if err != nil {
		return nil, err
	}

	return &Pass1Result{
		Macros:  macros,
		Symbols: symbols,
		Tokens:  expanded,
		PC:      pc,
	}, nil
}

func Assemble(source string) ([]int, error) {
	res, err := Pass1(source)
	if err != nil {
		return nil, err
	}

	var binary []int
	for _, t := range res.Tokens {
		if t.Type == "SYMBOL" {
			if addr, ok := res.Symbols[t.Value]; ok {
				binary = append(binary, addr)
			} else {
				return nil, fmt.Errorf("undefined label: %s", t.Value)
			}
		} else if t.Type == "NUMBER" {
			val, _ := strconv.ParseFloat(t.Value, 64)
			binary = append(binary, int(val))
		}
	}

	return binary, nil
}
